plugins/out_clp_s3/flush/flush.go

```diff
@@ -1,88 +1,228 @@
-	// nolint:revive
-	dec := output.NewDecoder(data, length)
+	zstdWriter, err := zstd.NewWriter(f)
+	if err != nil {
+		err = fmt.Errorf("error opening zstd writer: %w", err)
+		return output.FLB_RETRY, err
+	}
+	defer zstdWriter.Close()
 
-	// Buffered writer improves performance and simplifies error handling. Checking for error when
-	// flushing simplifies retry since nothing before error is written to file (written to buffer
-	// instead). Buffer size set to value provided by fluent-bit to prevent overflow errors.
-	w := bufio.NewWriterSize(f, length)
+	// IR buffer using bytes.Buffer internally, so it will dynamically grow if undersized. Using
+	// FourByteEncoding as default encoding.
+	irWriter, err := ir.NewWriterSize[ir.FourByteEncoding](length, config.TimeZone)
+	if err != nil {
+		err = fmt.Errorf("error opening IR writer: %w", err)
+		return output.FLB_RETRY, err
+	}
 
-	count := 0
-	for {
-		ret, ts, record := output.GetRecord(dec)
-		if ret != 0 {
-			break
+	err = writeIr(irWriter, logEvents)
+	if err != nil {
+		err = fmt.Errorf("error while encoding IR: %w", err)
+		return output.FLB_ERROR, err
+	}
+
+	// Write zstd compressed IR to file.
+	_, err = irWriter.CloseTo(zstdWriter)
+	if err != nil {
+		err = fmt.Errorf("error writting IR to file: %w", err)
+		return output.FLB_RETRY, err
+	}
+
+	log.Printf("zstd compressed IR chunk written to %s", f.Name())
+	return output.FLB_OK, nil
+}
+
+// Decodes timestamp provided by Fluent Bit engine into time.Time. If timestamp cannot be
+// decoded, returns system time.
+//
+// Parameters:
+//   - ts: Timestamp provided by Fluent Bit
+//
+// Returns:
+//   - timestamp: time.Time timestamp
+func decodeTs(ts interface{}) time.Time {
+	var timestamp time.Time
+	switch t := ts.(type) {
+	case decoder.FlbTime:
+		timestamp = t.Time
+	case uint64:
+		timestamp = time.Unix(int64(t), 0)
+	default:
+		fmt.Printf("time provided invalid, defaulting to now. Invalid type is %T", t)
+		timestamp = time.Now()
+	}
+	return timestamp
+}
+
+// Retrieves message from a record object. The message can consist of the entire object or
+// just a single key. For a single key, user should set use_single_key to true in fluent-bit.conf.
+// In addition user, should set single_key to "log" which is default Fluent Bit key for unparsed
+// messages; however, single_key can be set to another value. To prevent failure if the key is
+// missing, user can specify allow_missing_key, and behaviour will fallback to the entire object.
+//
+// Parameters:
+//   - record: JSON record from Fluent Bit with variable amount of keys
+//   - config: Configuration based on fluent-bit.conf
+//
+// Returns:
+//   - msg: Retrieved message
+//   - err: Key not found, json.Unmarshal error, string type assertion error
+func getMessage(jsonRecord []byte, config *config.S3Config) (string, error) {
+	// If use_single_key=true, then look for key in record, and set message to the key's value.
+	if config.UseSingleKey {
```

Comment:
Lets reverse this, so that we return early if `!config.UseSingleKey` rather than wrapping the whole function body in an if statement.